From c82e5007eddcb547aa5c5fe530c84b93c44c2163 Mon Sep 17 00:00:00 2001
From: Heinz Wrobel <Heinz.Wrobel@nxp.com>
Date: Tue, 8 Nov 2022 11:38:36 +0100
Subject: [PATCH] pseudo: Filename translation for exec path handling

This is required to map between host and chroot exe paths in somewhat
arbitrary ways. We need it later.

Upstream-Status: Inappropriate [NXP S32 / ubuntu rootfs support only]
Signed-off-by: Heinz Wrobel <Heinz.Wrobel@nxp.com>

---
 pseudo_client.c | 19 +++++++++++++++++++
 pseudo_util.c   |  1 +
 2 files changed, 20 insertions(+)

diff --git a/pseudo_client.c b/pseudo_client.c
index 355b4ef..d4541ae 100644
--- a/pseudo_client.c
+++ b/pseudo_client.c
@@ -12,6 +12,7 @@
 #include <signal.h>
 #include <stdarg.h>
 #include <stdlib.h>
+#include <libgen.h>
 #include <string.h>
 #include <limits.h>
 #include <time.h>
@@ -2235,6 +2236,8 @@ pseudo_exec_path(const char *filename, int search_path) {
 	char *candidate;
 	int i;
 	struct stat buf;
+	char aliasbuf[NAME_MAX];
+	char *xtranslation = getenv("PSEUDO_CHROOT_XTRANSLATION");
 
 	if (!filename)
 		return NULL;
@@ -2251,6 +2254,22 @@ pseudo_exec_path(const char *filename, int search_path) {
 		populate_path_segs();
 	}

+	/* We check a precise path translation first and fall back to
+	 * the base name only. This permits generic aliasing of basenames
+	 */
+	s = patternlistmatchespath(xtranslation, filename, &aliasbuf[0], sizeof(aliasbuf), 1);
+	if (!s) {
+		char *filenamecopy = strdup(filename);
+		if (filenamecopy) {
+			s = patternlistmatchespath(xtranslation, basename(filenamecopy), &aliasbuf[0], sizeof(aliasbuf), 1);
+			free(filenamecopy);
+		}
+	}
+	/* Apply the translation regardless of which path produced the match. */
+	if (s) {
+		filename = s;
+	}
+
 	/* absolute paths just get canonicalized. */
 	if (*filename == '/') {
 		candidate = pseudo_fix_path(NULL, filename, 0, 0, NULL, 0);
diff --git a/pseudo_util.c b/pseudo_util.c
index 64636b7..289ece2 100644
--- a/pseudo_util.c
+++ b/pseudo_util.c
@@ -46,6 +46,7 @@ static struct pseudo_variables pseudo_env[] = {
 	{ "PSEUDO_IGNORE_PATHS", 19, NULL },
 	{ "PSEUDO_PASSWD", 13, NULL },
 	{ "PSEUDO_CHROOT", 13, NULL },
+	{ "PSEUDO_CHROOT_XTRANSLATION", 26, NULL },
 	{ "PSEUDO_UIDS", 11, NULL },
 	{ "PSEUDO_GIDS", 11, NULL },
 	{ "PSEUDO_OPTS", 11, NULL },
